### ISSUE TITLE
Custom File and Line Overwriting

### ISSUE THREAD

**Issue by gluax** [tokio-rs/tracing#2419]
## Feature Request

This may already exist, but I could not find any documentation or examples on it, but please feel free to point me in the right direction if I'm wrong.

### Crates

N/A

### Motivation

I've been working with wasm a lot lately, specifically using the `wasmer` and `wasmer-wasi`. I have a use case where the logging function is written from the rust side. Then that function is called on the wasm side. This results in always saying the file and line are from the rust side. I'd love to be able to overwrite the level and line easily just by passing it to a macro.

### Proposal

Something like `tracing::debug!(message, line = 32, file = "wasm_bins/foo/main.rs")` would be great.

### Alternatives

I'm not super familiar with the internals of tracing; there may be better ways to do this.

**Comment by Ararem** [comment]
I'm not aware of a way to do this in the exact manner you intend, as in overwriting the `backtrace`/`spantrace` of your logs. However, you can attach *a custom field* to do a similar purpose, though it will require more work on your side as you will have to look at that information yourself and ignore what `tracing` tells you.

Events and spans can have extra information attached, called **fields**. These are essentially a name and a value to go along with it. See [the tracing docs](https://docs.rs/tracing/latest/tracing/index.html#using-the-macros) for more info on how to attach fields to spans/events. Essentially, for a `span` do 
```rust
debug_span!("this is a span", my_field_name=my_field_value, /*etc*/)
```
Events are the same:
```rust
trace!(
    field_1="MWAHAHAHAHA", my_second_field=%true /* % is for std::fmt::Display*/, debug_field=?a_debuggable_struct /*std::fmt::Debug*/,
    "MY EVENT MESSAGE: {format_arg:^-4} and {another}", /*Same syntax as `format!()` macro*/
    format_arg=3.5f32, another="hi!" /* `format!()` syntax again; the values you want to format into the string above */
)
```
You could do essentially what you had, just the other way around for events. The macros unfortunately don't allow you to overwrite the metadata AFAIK, just `target`, `parent` the message, and fields.

If you want a concrete example of how to use the macros, here's one copied from my code:
```rust
    let span_window = trace_span!(
        target: UI_TRACE_BUILD_INTERFACE,
        "build_window",
        window = label
    )
    .entered();
```

You *may* be able to directly create the span without the macros, but I have no idea how you'd go about that and don't really reccommend it.

**Comment by gluax** [comment]
Thanks, I currently have it set up this way!

I would still love to see the ability to overwrite line and file metadata at some point for the sake of ease and the same formatting.

**Comment by Ararem** [comment]
Yeah that would be nice. I've been doing some exploration and I now believe that `tracing` uses the backtrace to do this, so you might have to see if you can insert a custom stack frame or something, otherwise I think you're out of luck

:disappointed:

**Comment by hawkw** [comment]
`tracing` does not use backtraces at runtime. Instead, the file, line number, and module path are determined by the `tracing` macros using the `file!()`, `line!()`, and `module!()` macros: https://github.com/tokio-rs/tracing/blob/58accc6da3f04af3f6144fbe6d68af7225c70c02/tracing-core/src/lib.rs#L262-L271

It would probably be possible to change the macros to accept an overridden file, line number, and/or module path. However, because these values are part of a span or event's _metadata_ (which is determined at compile time), rather than its data (which is determined at runtime), they would have to be constants rather than variables. This may not work in your use case, if the same `tracing` macro invocation can correspond to multiple source code locations in a WASM module.

An alternative would be to rely on the support that some `tracing` subscribers, such as `tracing_subscriber::fmt`, provide for compatibility with the `log` crate. If a subscriber supports this, it may handle the event fields `log.file`, `log.line`, and `log.module_path` by treating those fields values as the file name, line number, and module path when displaying the formatted representation of that event. You could use those fields to override the displayed source locations for events coming from WASM. However, note that **these fields are recorded as data, and not as metadata**, so you won't be able to benefit from static filter caching with these events.
